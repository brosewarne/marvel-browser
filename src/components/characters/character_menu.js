import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { characterSearch } from '../../api/api';
import { CharacterTile } from './character_tile';
import { CharacterGrid } from './character_grid';
import { PaginationMenu } from '../menus/pagination_menu';
import { AlphabetMenu } from '../menus/alphabet_menu';

import './character.css';
import loading from './loading.svg';

class CharacterMenu extends Component {
    static propTypes = {
        match: PropTypes.object.isRequired,
        dispatch: PropTypes.func.isRequired
    };

    constructor(props) {
        super(props);

        this.state = {
            loading: true,
            startsWith: 'A',
            characters: [],
            // totalCharacters: 0,
            charactersPerPage: 9,
            currentPage: 1,
            totalPages: 0
        };
    }

    /**
     * Get the data for the set of characters based on `startsWith`, `currentPage` and `charactersPerPage` and set
     * the character list and total characters to internal state
     */
    getCharacterData = () => {
        const { startsWith, charactersPerPage, currentPage } = this.state;
        const offset = charactersPerPage * (currentPage - 1);
        const query = `nameStartsWith=${startsWith}&limit=${charactersPerPage}&offset=${offset}`;
        // @todo handle AJAX request errors here
        characterSearch(query).then((response) => {
            this.setState({
                loading: false,
                characters: response.data.results,
                // totalCharacters: response.data.total,
                totalPages: Math.ceil(response.data.total / this.state.charactersPerPage)
            });
        });
    };

    /**
     * Get the initial set of character data when the page mounts.
     */
    componentWillMount = () => {
        this.setState({ ...this.props.match.params }, this.getCharacterData);
    };

    /**
     * Update the set of character data when the route changes
     * @param {Object} nextProps - The incoming set of props
     */
    componentWillReceiveProps = (nextProps) => {
        this.setState({ ...nextProps.match.params }, this.getCharacterData);
    };

    /**
     * Get the pagination menu for the current set of characters
     * @returns {Node} - The rendered PaginationMenu
     */
    getPaginationMenu = () => {
        const { totalPages, startsWith } = this.state;
        return (
            <PaginationMenu
                totalPages={totalPages}
                startsWith={startsWith}
            />
        );
    };

    /**
     * Get the pagination details for display - eg Page 1 of 10
     * @returns {Node} - The pagination details
     */
    getPaginationDetails = () => {
        return (
            <strong>Page {this.state.currentPage} of {this.state.totalPages}</strong>
        );
    };

    /**
     * Return an alphabet menu for viewing characters that start with each letter of the alphabet
     * @returns {Node} - The rendered AlphabetMenu
     */
    getAlphabetMenu = () => {
        return <AlphabetMenu dispatch={this.props.dispatch} />;
    };

    /**
     * Create a set of character tiles for the current set of characters.
     * @returns {Array} - The array of rendered CharacterTiles
     */
    getCharacterTiles = () => {
        return this.state.characters.map((character) => {
            return <CharacterTile character={character} key={`character_tile_${character.id}`} />;
        });
    };

    /**
     * Create a loading Image with text.
     * @returns {Node} - The loading image and text
     */
    getLoadingImage = () => {
        return (
            <div>
                <h1>
                    LOADING CHARACTERS.....
                </h1>
                <div>
                    <img src={loading} className="loadingSpinner" alt="loading" />
                </div>
            </div>
        );
    };

    /**
     * Create a grid of character tiles for the current set of characters.
     * @returns {Node} - The character grid
     */
    getCharacterGrid = () => {
        if (this.state.loading) {
            return this.getLoadingImage();
        }
        const characterTiles = this.getCharacterTiles();
        return <CharacterGrid characterTiles={characterTiles} />;
    };

    render() {
        return (
            <div>
                <section>
                    {this.getAlphabetMenu()}
                </section>
                <content>
                    {this.getPaginationDetails()}
                    {this.getCharacterGrid()}
                    {this.getPaginationMenu()}
                </content>
            </div>
        );
    }
}

export { CharacterMenu };
