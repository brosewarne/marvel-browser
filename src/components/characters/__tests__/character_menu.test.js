import React from 'react';
import { shallow, mount, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import { CharacterMenu } from '../character_menu';

configure({ adapter: new Adapter() });

const createMockCharacters = (numCharacters) => {
    const mocks = [];
    for (let i = 0; i < numCharacters; i++) {
        mocks.push({
            id: i,
            name: `Character_${i}`,
            thumbnail: {
                path: '/a/href/to/an/image',
                extension: 'png'
            }
        });
    }
    return mocks;
};


describe('<CharacterMenu />', () => {
    // no need to use real CharacterTitle components here.
    const mockCharacterTiles = [
        <span key="c1">Character 1</span>,
        <span key="c2">Character 2</span>,
        <span key="c3">Character 3</span>,
        <span key="c4">Character 4</span>,
        <span key="c5">Character 5</span>,
        <span key="c6">Character 6</span>,
        <span key="c7">Character 7</span>,
        <span key="c8">Character 8</span>,
        <span key="c9">Character 9</span>,
        <span key="c10">Character 10</span>
    ];

    it('renders the loading spinner correctly', () => {
        const mockDispatch = jest.fn();
        const match = {
            params: {
                currentPage: 1,
                startsWith: 'A'
            }
        };
        const wrapper = shallow(
            <CharacterMenu dispatch={mockDispatch} match={match} />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders the loaded page correctly', () => {
        const mockDispatch = jest.fn();
        const match = {
            params: {
                currentPage: 1,
                startsWith: 'A'
            }
        };
        const mockCharacters = createMockCharacters(10);

        const wrapper = shallow(
            <CharacterMenu dispatch={mockDispatch} match={match} />
        );
        wrapper.setState({ loading: false, totalPages: 10, characters: mockCharacters });
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    // it('renders 4 character rows', () => {
    //     const wrapper = mount(
    //         <CharacterMenu characterTiles={mockCharacterTiles} />
    //     );
    //     expect(wrapper.find('.characterRow')).toHaveLength(4);
    // });
    //
    // it('renders 3 characters per row until the last row', () => {
    //     const wrapper = mount(
    //         <CharacterMenu characterTiles={mockCharacterTiles} />
    //     );
    //     expect(wrapper.find('.characterRow').first().children()).toHaveLength(3);
    //     expect(wrapper.find('.characterRow').last().children()).toHaveLength(1);
    // });
});
