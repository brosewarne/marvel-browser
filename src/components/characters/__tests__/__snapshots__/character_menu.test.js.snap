// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterMenu /> renders the loaded page correctly 1`] = `
<div>
  <section>
    <Component
      dispatch={[MockFunction]}
    />
  </section>
  <content>
    <strong>
      Page 
      1
       of 
      10
    </strong>
    <Component
      characterTiles={
        Array [
          <Unknown
            character={
              Object {
                "id": 0,
                "name": "Character_0",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 1,
                "name": "Character_1",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 2,
                "name": "Character_2",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 3,
                "name": "Character_3",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 4,
                "name": "Character_4",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 5,
                "name": "Character_5",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 6,
                "name": "Character_6",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 7,
                "name": "Character_7",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 8,
                "name": "Character_8",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
          <Unknown
            character={
              Object {
                "id": 9,
                "name": "Character_9",
                "thumbnail": Object {
                  "extension": "png",
                  "path": "/a/href/to/an/image",
                },
              }
            }
          />,
        ]
      }
    />
    <Component
      startsWith="A"
      totalPages={10}
    />
  </content>
</div>
`;

exports[`<CharacterMenu /> renders the loading spinner correctly 1`] = `
<div>
  <section>
    <Component
      dispatch={[MockFunction]}
    />
  </section>
  <content>
    <strong>
      Page 
      1
       of 
      0
    </strong>
    <div>
      <h1>
        LOADING CHARACTERS.....
      </h1>
      <div>
        <img
          alt="loading"
          className="loadingSpinner"
          src="loading.svg"
        />
      </div>
    </div>
    <Component
      startsWith="A"
      totalPages={0}
    />
  </content>
</div>
`;
